### ISSUE TITLE
Python virtualenv: undefined method `delete'

### ISSUE THREAD

**Issue by JarryShaw** [Homebrew/brew#7896]
## What you were trying to do (and why)

So, as I recently noticed, for a formula based on Python (see below), `brew install` will fail with `NoMethodError`, claiming that ``undefined method `delete' for nil:NilClass``.

```ruby
class Foo < Formula
  include Language::Python::Virtualenv

  ...

  # when using "python@3.8" in a Tap, `brew audit` fails with 
  #   Dependency 'python@3.8' from homebrew/core is an alias; use the canonical name 'python@3.8'.
  # depends_on "python@3.8"
  depends_on "homebrew/core/python@3.8"

  ...

  def install
    virtualenv_install_with_resources

    ...
  end

  ...
end
```

## What happened (include command output)

`brew install foo` fails at `virtualenv_install_with_resources` with `NoMethodError`, with backtrace as below.

```
Error: An exception occurred within a child process:
  NoMethodError: undefined method `delete' for nil:NilClass
/usr/local/Homebrew/Library/Homebrew/language/python.rb:190:in `virtualenv_install_with_resources'
/usr/local/Homebrew/Library/Taps/jarryshaw/homebrew-tap/Formula/walrus.rb:24:in `install'
/usr/local/Homebrew/Library/Homebrew/debrew.rb:24:in `block in install'
/usr/local/Homebrew/Library/Homebrew/debrew.rb:94:in `debrew'
/usr/local/Homebrew/Library/Homebrew/debrew.rb:24:in `install'
/usr/local/Homebrew/Library/Homebrew/build.rb:152:in `block (2 levels) in install'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1176:in `block in brew'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2171:in `block (2 levels) in stage'
/usr/local/Homebrew/Library/Homebrew/utils.rb:474:in `with_env'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2170:in `block in stage'
/usr/local/Homebrew/Library/Homebrew/resource.rb:111:in `block in unpack'
/usr/local/Homebrew/Library/Homebrew/resource.rb:197:in `block in mktemp'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:57:in `block in run'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:57:in `chdir'
/usr/local/Homebrew/Library/Homebrew/mktemp.rb:57:in `run'
/usr/local/Homebrew/Library/Homebrew/resource.rb:196:in `mktemp'
/usr/local/Homebrew/Library/Homebrew/resource.rb:106:in `unpack'
/usr/local/Homebrew/Library/Homebrew/resource.rb:81:in `stage'
/usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/lib/ruby/2.6.0/forwardable.rb:230:in `stage'
/usr/local/Homebrew/Library/Homebrew/formula.rb:2145:in `stage'
/usr/local/Homebrew/Library/Homebrew/formula.rb:1169:in `brew'
/usr/local/Homebrew/Library/Homebrew/build.rb:123:in `block in install'
/usr/local/Homebrew/Library/Homebrew/utils.rb:474:in `with_env'
/usr/local/Homebrew/Library/Homebrew/build.rb:118:in `install'
/usr/local/Homebrew/Library/Homebrew/build.rb:202:in `<main>'
```

## What you expected to happen

Successfully install the formula as using `depends_on "python@3.8"`.

## Step-by-step reproduction instructions (by running `brew` commands)

```
$ brew cat my/tap/foo  # just a dummy formula
class Foo < Formula
  include Language::Python::Virtualenv
  ...
  depends_on "homebrew/core/python@3.8"
  ...
  def install
    virtualenv_install_with_resources
    ...
  end
  ...
end
$ brew install my/tap/foo
...
Error: An exception occurred within a child process:
  NoMethodError: undefined method `delete' for nil:NilClass
```

## Output of `brew config` and `brew doctor` commands

<pre>
$ brew config
HOMEBREW_VERSION: 2.4.3-31-gf1e4c4e
ORIGIN: https://github.com/Homebrew/brew
HEAD: f1e4c4ed56a9f719534b8db04af13a8c4617c537
Last commit: 4 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 4a39d1501103b581a935b83117e0b9f0f1fbf7c7
Core tap last commit: 2 hours ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_BOTTLE_DOMAIN: https://mirrors.tuna.tsinghua.edu.cn/homebrew-bottles
HOMEBREW_EDITOR: subl
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 12
HOMEBREW_NO_BOTTLE_SOURCE_FALLBACK: set
CPU: dodeca-core 64-bit kabylake
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
Clang: 11.0 build 1103
Git: 2.27.0 => /usr/local/bin/git
Curl: 7.64.1 => /usr/bin/curl
Java: 1.8.0_252
macOS: 10.15.6-x86_64
CLT: 1103.0.32.62
Xcode: 11.5
$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: "config" scripts exist outside your system or Homebrew directories.
`./configure` scripts often look for *-config scripts to determine if
software packages are installed, and which additional flags to use when
compiling and linking.

Having additional scripts in your path can confuse software installed via
Homebrew if the config script overrides a system or Homebrew-provided
script of the same name. We found the following "config" scripts:
  /Users/cestmoi/.pyenv/shims/python2-config
  /Users/cestmoi/.pyenv/shims/python-config
  /Users/cestmoi/.pyenv/shims/python3-config
  ...  # other pyenv shipped python-config executives
</pre>

**Comment by MikeMcQuaid** [comment]
> /usr/local/Homebrew/Library/Taps/jarryshaw/homebrew-tap/Formula/walrus.rb:24:in `install'

Can you provide the full formula definition here? Thanks!

**Comment by JarryShaw** [comment]
> > /usr/local/Homebrew/Library/Taps/jarryshaw/homebrew-tap/Formula/walrus.rb:24:in `install'
> 
> Can you provide the full formula definition here? Thanks!

Voila (from `jarryshaw/tap/walrus`):

```ruby
class Walrus < Formula
  include Language::Python::Virtualenv

  desc "Backport compiler for Python 3.8 assignment expressions"
  homepage "https://github.com/pybpc/walrus#walrus"
  url "https://github.com/pybpc/walrus/archive/v0.1.4.tar.gz"
  sha256 "9c568b24f33f38bad9625ca38f8d1cd3cef479e07951b4f071d30dafd967c0ae"

  head "https://github.com/pybpc/walrus.git", :branch => "master"

  depends_on "homebrew/core/python@3.8"

  resource "parso" do
    url "https://files.pythonhosted.org/packages/e0/a2/3786c568fc8e9f64b9f7143e4c3904e61a8c5cb17260f22a6a3019d80c02/parso-0.5.2.tar.gz"
    sha256 "55cf25df1a35fd88b878715874d2c4dc1ad3f0eebd1e0266a67e1f55efccfbe1"
  end

  resource "tbtrim" do
    url "https://files.pythonhosted.org/packages/85/62/89756f5d2d61691591c4293fd4cc1fbb3aab1466251c7319fe60dd08fb27/tbtrim-0.3.1.tar.gz"
    sha256 "b6285ac02e9a7b78fab97de65668fe2def0f5d8783b0e0dfcb4b7c0a635b3d11"
  end

  def install
    virtualenv_install_with_resources

    man1.install "share/walrus.1"
    bash_completion.install "share/walrus.bash-completion"
  end

  test do
    (testpath/"test.py").write <<~EOS
      a = (b := 1)
      print(a, b)
    EOS

    std_output = <<~EOS
      1 1
    EOS

    system bin/"walrus", "--no-archive", "test.py"
    assert_match std_output, shell_output("#{Formula["python3"].bin}/python test.py")
  end
end
```

And I also tried this on formulae in the `homebrew/core`, e.g. `legit`, the same ``NoMethodError: undefined method `delete' for nil:NilClass`` exception occurred when I changed `depends_on "python@3.8"` to `depends_on "homebrew/core/python@3.8"`.

**Comment by Rylan12** [comment]
It looks like the issue is due to the `depends_on "homebrew/core/python@3.8"`.

`virtualenv_install_with_resources` tries to determine which `python` by using the `needs_python?` method:

https://github.com/Homebrew/brew/blob/505c1e72405840ca0937420e2aae5cc70b01b7da/Library/Homebrew/language/python.rb#L167-L171

In this case, `python` is any of the following: `python`, `python3`, `python@3`, `python@3.8`, `pypy`, `pypy3` (this is determined in `virtualenv_install_with_resources`.

`needs_python?` looks at the dependencies of the formula to see if any match:

https://github.com/Homebrew/brew/blob/505c1e72405840ca0937420e2aae5cc70b01b7da/Library/Homebrew/language/python.rb#L170

In this case, `r.name` is `homebrew/core/python@3.8` which does not equal any of the options above. In homebrew core, `r.name` doesn't include the `homebrew/core/` prefix so there isn't an issue here.

@MikeMcQuaid I'm not sure what the right solution here is as I haven't worked with taps using homebrew-core formulae. One solution I can think of is to compare only the basename of `r.name` (e.g. `python@3.8` instead of `homebrew/core/python@3.8`).

**Comment by Bo98** [comment]
`r.name.split("/").last` seems appropriate - assuming we want to also allow `some/tap/python`. We do the last split thing in some other places in `brew`.

**Comment by Rylan12** [comment]
Fix in #7899
